### PR TITLE
Add profile recorder and refactor day of year and week indices.

### PR DESCRIPTION
**Pull request by jetuk**
A new recorder for storing a calculated daily profile. Useful for optimisation where a new profile might be calculated from non-daily profiles and/or combinations of parameters.

### CODE REVIEW

File: pywr/recorders/_recorders.pyx [hunk 1]
```diff
@@ -1278,6 +1278,84 @@ cdef class NumpyArrayParameterRecorder(ParameterRecorder):
 NumpyArrayParameterRecorder.register()
 
 
+cdef inline bint is_leap_year(int year):
```
[the remark this answers is not in the document]
Comment (jetuk) [reply pywr/recorders/_recorders.pyx:1281]: Not if it is inline.

File: pywr/recorders/_recorders.pyx [hunk 2]
```diff
@@ -1278,6 +1278,84 @@ cdef class NumpyArrayParameterRecorder(ParameterRecorder):
...
+    param : `pywr.parameters.Parameter`
+        Parameter instance to record.
+    temporal_agg_func : str or callable (default="mean")
+        Aggregation function used over time when computing a value per scenario. For aggregation over scenarios
+        see the `agg_func` keyword argument.
+    """
+    def __init__(self, model, Parameter param, **kwargs):
+        # Optional different method for aggregating across time.
+        temporal_agg_func = kwargs.pop('temporal_agg_func', 'mean')
+        super(NumpyArrayDailyProfileParameterRecorder, self).__init__(model, param, **kwargs)
```
[the remark this answers is not in the document]
Comment (jetuk) [reply pywr/recorders/_recorders.pyx]: Done.

File: pywr/recorders/_recorders.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (jetuk) [reply pywr/recorders/_recorders.pyx]: I've refactored everything to do with leap years and days of year indexes. I think we spoke about doing this a while ago.

[the remark this answers is not in the document]
Comment (jetuk) [reply pywr/recorders/_recorders.pyx]: The issue is #804

**Comment by jetuk** [comment]
This now entirely fixes #804.

**Comment by jetuk** [comment]
My only question is whether it should be `Timestep.week_index` instead of `Timestep.week` to be consistent with the `.dayofyear_index` which is also zero-based.

**Comment by Batch21** [comment]
> My only question is whether it should be `Timestep.week_index` instead of `Timestep.week` to be consistent with the `.dayofyear_index` which is also zero-based.

I've a slight preference for  `Timestep.week_index` and I think it leaves less scope for confusion. Otherwise changes looks good

**Comment by jetuk** [comment]
> > My only question is whether it should be `Timestep.week_index` instead of `Timestep.week` to be consistent with the `.dayofyear_index` which is also zero-based.
> 
> I've a slight preference for `Timestep.week_index` and I think it leaves less scope for confusion. Otherwise changes looks good

Done.

**Pull request merged**